src/java.desktop/share/classes/javax/swing/ImageIcon.java

```diff
@@ -436,10 +439,28 @@ public void setDescription(String description) {
      * @param y the Y coordinate of the icon's top-left corner
      */
     public synchronized void paintIcon(Component c, Graphics g, int x, int y) {
-        if(imageObserver == null) {
-           g.drawImage(image, x, y, c);
-        } else {
-           g.drawImage(image, x, y, imageObserver);
+        Object oldHint = null;
+        boolean hintChanged = false;
+        ImageObserver observer = imageObserver != null ? imageObserver : c;
+        if (image instanceof MultiResolutionImage) {
+            Image variant = ((MultiResolutionImage) image).getResolutionVariant(this.width, this.height);
+            if (variant.getHeight(observer) != height || variant.getWidth(observer) != width) {
+                if (g instanceof Graphics2D) {
+                    Graphics2D g2d = (Graphics2D) g;
+                    RenderingHints rh = g2d.getRenderingHints();
+                    oldHint = rh.get(RenderingHints.KEY_INTERPOLATION);
+                    if (!RenderingHints.VALUE_INTERPOLATION_BICUBIC.equals(oldHint)) {
+                        g2d.setRenderingHint(RenderingHints.KEY_INTERPOLATION,
+                                RenderingHints.VALUE_INTERPOLATION_BICUBIC);
+                        hintChanged = true;
+                    }
+                }
+            }
+        }
+        g.drawImage(image, x, y, observer);
+        if (hintChanged) {
+            ((Graphics2D) g).setRenderingHint(RenderingHints.KEY_INTERPOLATION,
+                    oldHint == null ? RenderingHints.VALUE_INTERPOLATION_NEAREST_NEIGHBOR : oldHint);
```

Comment:
Isn't the state of the Graphics object modified? Previously, the `KEY_INTERPOLATION` hint was set to `null` but it is set to `VALUE_INTERPOLATION_NEAREST_NEIGHBOR` on exit. Why can't it be set to `null`? Always to `oldHint`?